### PR TITLE
NetSim: Send a message v1

### PR DESCRIPTION
**Pull request by islemaster**
This PR introduces the ability to send and receive messages between NetSim clients, via a simulated router.

screenshot from 2015-02-09 14 29 28

New concepts in this PR:
- `NetSimMessage`, a new kind of entity that lives in its own '_message' table.  This represents a message "on the wire" in transit from one node to another - the first thing the receiving node should do when processing a message is delete its row.  Messages always have the structure `{fromNodeID:, toNodeID:, payload:}` where the node IDs are metadata and the payload contains the actual information on the wire, including from and to addresses.
- `NetSimLogWidget` which for now is a simple clearable scrollback box, but will eventually have more features.
- `NetSimSendWidget` likewise is a very simple form for now.
- The `NetSimTable` type in NetSimShard.js, which wraps the app storage table API in a local cache layer and provides a notification-like interface to the rest of NetSim.  I'm only using this for the messages table now, but as cleanup will be retrofitting the lobby widget and router with this functionality.  There's some fairly heavy collection processing in there right now.  My hope is when we get real notifications hooked up this can go away - or at least be a backup that doesn't happen too often.

Here's how sending a message works:
1. The user enters a "To" address and a message.  The send widget packages this into a packet in the form `{ toAddress: 1, messageContent: "Hello!" }` and passes it to `myNode.sendMessage()`
2. myNode looks at its single wire to determine the source and destination Node IDs, and creates a message in the '_messages' table.
3. Creating the message immediately triggers a table change event on the local browser, which the local router simulation receives.  It scans the new table contents for any messages from the local client's Node ID to the router's Node ID.
4. Finding a new message, the router copies its contents and deletes it from the table.  It then examines the payload for a `toAddress`.  If it finds one, it does a lookup against its own set of wires for a matching address.  There are a few ways the message can get swallowed:
   - No `toAddress` field: Malformed packet, dropped.
   - The `toAddress` is zero, the router's address.  Packet dropped.
   - The `toAddress` is not on the local network.  Packet should be forwarded out of the network, but we don't do that yet, so packet dropped.
5. Once a matching wire is found, the router creates a **new** message with new to and from Node IDs (from the router to  the destination) but the same payload.
6. The next time the remote client table polls for updates (5 seconds for now) it will report a table change event to the client node simulation, which scans for packets to its Node ID.  Finding one, it pulls it from the wire (deletes the row) and passes the payload to the received message log widget.

### CODE REVIEW

File: apps/src/netsim/NetSimLogWidget.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * Static counter used to generate/uniquely identify different instances
+ * of this log widget on the page.
+ * @type {number}
+ */
+NetSimLogWidget.uniqueIDCounter = 0;
+
+/**
+ * Generate a new NetSimLogWidget, puttig it on the page and hooking
```
Comment (davidsbailey) [inline apps/src/netsim/NetSimLogWidget.js]: s/puttig/putting

File: apps/src/netsim/NetSimLogWidget.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ * @param {!string} title - The log widget header text
+ */
+NetSimLogWidget.createWithin = function (element, title) {
+  var controller = new NetSimLogWidget();
+
+  var instanceID = NetSimLogWidget.uniqueIDCounter;
+  NetSimLogWidget.uniqueIDCounter++;
+
+  element.innerHTML = markup({
+    logInstanceID: instanceID,
```
Comment (davidsbailey) [inline apps/src/netsim/NetSimLogWidget.js]: Are we calling "instance" "shard" now, or is this different? sorry to harp on this, I'm not actually allergic to the word "instance"

**Comment by davidsbailey** [comment]
question regarding PR description part 4: why does the router have address zero, and how is dropping a packet addressed to 0 any different from dropping a packet addressed to any other number which is not on the local network?

File: apps/src/netsim/NetSimLogWidget.js (hunk 2 above)
Comment (islemaster) [inline apps/src/netsim/NetSimLogWidget.js]: Yes, I've swapped in "shard" everywhere so we're no longer mentally overloading the word "instance."  A shard is private NetSim environment for a group of users.

**Comment by islemaster** [comment]
I believe the suggested outcome is that the router has address zero, but we hide it from the address table.  There is no difference in outcome between sending a packet addressed to the router and sending a packet addressed outside the network, for now.  I suppose it could respond to messages at some point, though.

File: apps/src/netsim/NetSimNodeRouter.js [hunk 3]
```diff
@@ -110,16 +126,16 @@ var RouterStatus = NetSimNodeRouter.RouterStatus;
 /**
  * Updates router status and lastPing time in lobby table - both keepAlive
  * and making sure router's connection count is valid.
- * @param onComplete
+ * @param {function} [onComplete] - Optional success/failure callback
```
Comment (davidsbailey) [inline apps/src/netsim/NetSimNodeRouter.js]: remove square brackets to match existing style in your other comments?

**Comment by davidsbailey** [comment]
LGTM after a quick look. If you have any specific concerns, let's go over them together in more detail.

File: apps/src/netsim/NetSimNodeRouter.js (hunk 3 above)
Comment (islemaster) [inline apps/src/netsim/NetSimNodeRouter.js]: Square brackets denote an [optional parameter in JSDoc](http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values).

**Pull request merged**